File: alica_engine/src/engine/modelmanagement/factories/QuantifierFactory.cpp
#include "engine/modelmanagement/factories/QuantifierFactory.h"
#include "engine/modelmanagement/Strings.h"
#include "engine/model/Quantifier.h"
#include "engine/model/ForallAgents.h"

namespace alica
{
    Quantifier* QuantifierFactory::create(const YAML::Node& quantifierNode)
    {
        Quantifier* quantifier;
        std::string quantifierType;
        if (Factory::isValid(quantifierNode[alica::Strings::quantifierType])) {
            quantifierType = Factory::getValue<std::string>(quantifierNode, alica::Strings::quantifierType);
        }
        if (quantifierType.compare(alica::Strings::forAllAgents) == 0) {
            quantifier = new ForallAgents();
        } else {
            AlicaEngine::abort("QuantifierFactory: Unknown quantifier type encountered!");
        }
        Factory::setAttributes(quantifierNode, quantifier);
        Factory::storeElement(quantifier, alica::Strings::quantifier);

        if (Factory::isValid(quantifierNode[alica::Strings::scope])) {
            Factory::quantifierScopeReferences.push_back(std::pair<int64_t, int64_t>(quantifier->getId(), Factory::getReferencedId(quantifierNode[alica::Strings::scope])));
        }

        if (Factory::isValid(quantifierNode[alica::Strings::sorts])) {
            const YAML::Node &sorts = quantifierNode[alica::Strings::sorts];
            for (YAML::const_iterator it = sorts.begin(); it != sorts.end(); ++it) {
                quantifier->_domainIdentifiers.push_back((*it).as<std::string>());
            }
        }

        return quantifier;
    }

    void QuantifierFactory::attachReferences() {
        // quantifierScopeReferences
        for (std::pair<int64_t, int64_t> pairs : Factory::quantifierScopeReferences) {
            AlicaElement* ael = (AlicaElement*) Factory::getElement(pairs.second);
            Quantifier* q = (Quantifier*) Factory::getElement(pairs.first);
            q->setScope(ael);
        }
        Factory::quantifierScopeReferences.clear();
    }
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/QuantifierFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.PlanElement;
import de.unikassel.vs.alica.planDesigner.alicamodel.Quantifier;
import de.unikassel.vs.alica.planDesigner.modelmanagement.Types;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import java.util.HashMap;

public class QuantifierFactory extends Factory {

    public static Quantifier create (Element quantifierNode, ConversionProcess cp) {
        Quantifier quantifier = new Quantifier();
        String quantifierType = quantifierNode.getAttribute(Tags.XSITYPE);
        if (quantifierType.equals(Tags.ALLAGENTSTAG)) {
            quantifier.setQuantifierType(Types.QUANTIFIER_FORALL);
        } else {
            throw new RuntimeException("[QuantifierFactory] Unknown quantifier type: '" + quantifierType +"'");
        }
        Factory.setAttributes(quantifierNode, quantifier);
        cp.addElement(quantifier);

        cp.quantifierScopeReferences.put (quantifier.getId(), cp.getReferencedId(quantifierNode.getAttribute(Tags.SCOPE)));

        NodeList sortNodeList = quantifierNode.getElementsByTagName(Tags.SORTS);
        for (int i = 0; i < sortNodeList.getLength(); i++) {
            Element sortNode = (Element) sortNodeList.item(i);
            quantifier.addSort(sortNode.getTextContent());
        }

        return quantifier;
    }

    public static void attachReferences(ConversionProcess cp) {
        for (HashMap.Entry<Long, Long> entry : cp.quantifierScopeReferences.entrySet()) {
            Quantifier quantifier = (Quantifier) cp.getElement(entry.getKey());
            PlanElement planElement = cp.getElement(entry.getValue());
            quantifier.setScope(planElement);
        }
        cp.quantifierScopeReferences.clear();
    }
}
